processor/src/kitspaceYaml.ts

```diff
@@ -0,0 +1,100 @@
+import { promises as fs } from 'fs'
+import jsYaml from 'js-yaml'
+import log from 'loglevel'
+import path from 'node:path'
+import { z } from 'zod'
+
+const eda = z.object({ type: z.enum(['kicad', 'eagle']), pcb: z.string() })
+
+export const singleKitspaceYaml = z.object({
+  summary: z.string().default(''),
+  site: z.string().url().or(z.literal('')).default(''),
+  color: z
+    .enum(['green', 'red', 'blue', 'black', 'white', 'orange', 'purple', 'yellow'])
+    .default('green'),
+  bom: z.string().optional(),
+  gerbers: z.string().optional(),
+  eda: eda.optional(),
+  readme: z.string().optional(),
+  'pcb-services': z
+    .enum(['aisler', 'pcbway', 'oshpark', 'jlcpcb'])
+    .array()
+    .optional(),
+  'ibom-enabled': z.boolean().default(true),
+  // a single kitspaceYaml doesn't have the multi key
+  multi: z.undefined(),
+})
+
+export type SingleKitspaceYaml = z.infer<typeof singleKitspaceYaml>
+
+const multiKey = z
+  .string()
+  .refine(key => key !== '_', { message: 'Cannot use "_" as project name.' })
+
+export const multiKitspaceYaml = z.object({
+  multi: z.record(multiKey, singleKitspaceYaml),
+})
+
+export type MultiKitspaceYaml = z.infer<typeof multiKitspaceYaml>
+
+export const kitspaceYamlInput = multiKitspaceYaml.or(singleKitspaceYaml)
+
+export type KitspaceYamlInput = z.infer<typeof kitspaceYamlInput>
+
+export const kitspaceYaml = singleKitspaceYaml.extend({ name: z.string() })
+
+export type KitspaceYaml = z.infer<typeof kitspaceYaml>
+
+export async function getKitspaceYaml(
+  inputDir: string,
+): Promise<Array<KitspaceYaml>> {
+  const filePaths = [
+    'kitspace.yaml',
+    'kitspace.yml',
+    'kitnic.yaml',
+    'kitnic.yml',
+  ].map(p => path.join(inputDir, p))
+  let input: KitspaceYamlInput
+  try {
+    const yamlContents = await Promise.all(filePaths.map(tryReadFile)).then(
+      ([yaml, yml, kitnicYaml, kitnicYml]) =>
+        yaml || yml || kitnicYaml || kitnicYml,
+    )
+    const obj = jsYaml.load(yamlContents)
+    input = kitspaceYamlInput.parse(obj)
+  } catch (e) {
+    log.warn('Could not parse Kitspace YAML file:', e)
+    // return the default
+    return [kitspaceYaml.parse({ name: '_' })]
+  }
+
+  if (!input.multi) {
+    return [{ ...input, name: '_' } as KitspaceYaml]
+  }
+
+  const arr: Array<KitspaceYaml> = []
+  for (const key of Object.keys(input.multi)) {
+    arr.push({ ...input.multi[key], name: formatAsGiteaRepoName(key) })
+  }
+  return arr
+}
```

Comment:
This section is a bit ambiguous, more documentation would be helpful!

Reply:
I've added some comments with [539c8c4](https://github.com/kitspace/kitspace-v2/pull/478/commits/539c8c4cc6732e73a2bd70dc6cab9053bf617370)